kpsaliency/datasets/saliencymap.py

```diff
@@ -0,0 +1,161 @@
+        # with open(params_path, "w") as f:
+        #     json.dump(bms_generator.params, f)
+        # print(f"Written parameters to {params_path}.")
+        print(f"Running with {n_workers} workers.")
+        total = len(self)
+        batch_size = total // n_workers
+        batch_idxs = [batch_size * n for n in range(n_workers)]
+        idxs = list(range(total))
+        batches = [idxs[n:n+batch_size] if n != batch_idxs[-1] else idxs[n:]
+                   for n in batch_idxs]
+        pbar_queue = mp.Queue()
+
+        desc = "Generating saliency map dataset"
+        pbar_proc = mp.Process(target=handle_pbar, args=(pbar_queue, total, desc))
+        pbar_proc.start()
+
+        processes = [mp.Process(target=SaliencyMapDataset._generate_dataset_batch,
+                                args=(self, batch, bms_generator, pbar_queue))
+                     for batch in batches]
+        for p in processes:
+            p.start()
+        for p in processes:
+            p.join()
+        pbar_proc.join()
+
+        self.load_images = _load_images
+        self.dataset_exists = True
+
+    def _generate_dataset_batch(self, idxs, bms_generator, pbar_queue: mp.Queue):
+        _bms_generator = bms_generator
+        for i in idxs:
+            img, info, vehicles = super().__getitem__(i)
+
+            if type(bms_generator) is dict:
+                _bms_generator = bms_generator[info.sequence.directory]
+
+            direct_kps = []
+            indirect_kps = []
+
+            for vehicle in vehicles:
+                for instance in vehicle.instances:
+                    if instance.direct:
+                        direct_kps.append(instance)
+                    else:
+                        indirect_kps.append(instance)
+
+            direct_smaps = _bms_generator.generate_cumulated_sm(img, direct_kps)
+
+            indirect_smaps = _bms_generator.generate_cumulated_sm(img, indirect_kps)
+
+            # convert maps to uint8 [0, 255]
+            direct_smaps = (direct_smaps * 255).astype(np.uint8)
+            indirect_smaps = (indirect_smaps * 255).astype(np.uint8)
+
+            # save maps
+            cv2.imwrite(
+                os.path.join(self.smap_path_direct, info.sequence.directory,
+                             info.file_name), direct_smaps)
+            cv2.imwrite(
+                os.path.join(self.smap_path_indirect, info.sequence.directory,
+                             info.file_name), indirect_smaps)
+
+            pbar_queue.put(1)
+
+    def __getitem__(self, idx):
+        img, info, vehicles = super().__getitem__(idx)
+        smap_direct = cv2.imread(os.path.join(
+            self.smap_path_direct, info.sequence.directory, info.file_name
+        ))
+        smap_indirect = cv2.imread(os.path.join(
+            self.smap_path_indirect, info.sequence.directory, info.file_name
+        ))
+
+        return img, smap_direct, smap_indirect, info, vehicles
```

Comment:
The getitem function returns a custom class with "info". Therefore it cannot be collated by the default collate function. This is by the way also the case for the getitem super function of the PVDNDataset class.

Reply:
The pvdn package already contains a custom `collate_fn` function.

Reply:
Are you sure about this? I do not remember this being the case. On which version (pvdn or pvdn_dev) and which branch can I see this collate function?

Reply:
lets discuss this in the pvdn repo.

Reply:
custom `collate_fn` is added, tutorial is adjusted accordingly